README.md
# screen_autosize

<p align="center">
   <a href="https://pub.flutter-io.cn/packages/screen_autosize">
    <img src="https://img.shields.io/pub/v/screen_autosize.svg" alt="pub package" />
  </a>
  <a href="http://www.apache.org/licenses/LICENSE-2.0">
    <img src="http://img.shields.io/badge/License-Apache%202.0-blue.svg?style=flat-square" alt="License" />
  </a>
  <a href="https://juejin.cn/user/3227821828236392">
    <img src="https://img.shields.io/badge/Author-CxmyDev-orange" alt="Author" />
  </a>
</p>


🔥 A low-cost Flutter screen adaptation solution (参考今日头条 Android 屏幕适配方案的实现原理，实现的一个极低成本的 Flutter 屏幕适配方案)。

Flutter 屏幕适配方案，可做到 100% 还原设计稿。


## Notice

- 使用方法
- 实现原理（挖坑后补）

## 优势 & 劣势

### 优势：

1. 适配效果：100% 还原 UI；
2. 稳定性高：原理简单，无任何 Framework 的 Hook；
3. 侵入性低：修改点少且统一，日常编写 UI 无需特殊注意；
4. 使用成本低：统一修改，全局有效；
5. 性能损耗：几乎没有；
6. 不破坏 const Widget 常量优化；

### 劣势：

1. 全局有效，无法单页面设置，需统一同项目不同模块的设计和开发标准；
2. 编码注意点：不能使用 window 类 Api 获取尺寸信息；
3. 需要重写 WidgetsFlutterBinding，可能与部分库存在冲突，需特殊处理；

## 使用方法

### 一、安装依赖

安装之前，请查看最新版本（新版本如有问题，请使用上一版本）。

```yaml
dependencies:
    flutter:
      sdk: flutter
    # 添加依赖
    screen_autosize: ^{latest version}
```

### 二、使用导包

```dart
import 'package:screen_autosize/screen_autosize.dart';
```

### 三、初始化

**Step1：设定基准屏幕宽度**

需要在 `runApp()` 之前设置，使用 AutoSizeUtils 的 initConfig() 设置设计稿的基准宽度。

```dart
void main() {
  // 这里使用 iPhone 一倍的宽度作为基准宽度；
  AutoSizeUtils.instance.initConfig(375);
  // ...
}
```

**Step2：替换 runApp()**

将原本的 `runApp()` 用 `runAutoSizeApp(MyApp())` 替换。

```dart
void main() {
  // 这里使用 iPhone 一倍的宽度作为基准宽度；
  AutoSizeUtils.instance.initConfig(375);
  // runApp(MyApp());
  runAutoSizeApp(MyApp());
}
```

**Step3: 替换 MaterialApp 生成的 MediaQuery**

MaterialApp 内部会生成 MediaQuery，需要将其通过 MediaQueryWrapper 替换。

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      builder: (context, widget) {
        // 替换根的 MediaQuery
        return MediaQueryWrapper(builder: (BuildContext context){
          return widget;
        },);
      },
      home: HomePage(title: 'Flutter ScreenAutoSize示例'),
    );
  }
}
// ...
```

### Step3: 开始写 UI

编码时，无任何注意点，直接按照设计图的尺寸写就行。

eg. 在 375 宽度（iPhone 1 倍尺寸）的设计稿下，一个 100x100 黄色区域，直接写参数就行，无需任何注意点。

```dart
class ColorsWidget extends StatelessWidget{
  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(10), // 不需要转换，设计稿是多少，就写多少
      width: 100,  // 不需要转换，设计稿是多少，就写多少
      height: 100, // 不需要转换，设计稿是多少，就写多少
      color: Colors.orange,
      child: Text(
        '我是正方形,边长是100',
        style: TextStyle(
          color: Colors.white,
          fontSize: 12,
        ),
      ),
    );
  }
}
```

如果要写满屏的尺寸，例如某个 Widget 宽度需要撑满屏幕，可以使用 2 种方法。

1. 利用 MediaQuery 获取：`MediaQuery.of(context).size.width`;
2. 利用 AutoSizeUtils 获取：`AutoSizeUtils.instance.mediaWidth`;

与宽度（width）对应的还有高度（height）的尺寸。

Tips：需要注意，因为库里改写了 devicePixelRatio，所以不能直接从 window 中获取对应参数。

## 效果

未适配前：从左到右的 dip 分别是 375 → 392 → 440

![old_size](https://raw.githubusercontent.com/CxmyDev/screen_autosize/main/art/old_size.jpg)

使用 screen_autosize 适配后：从左到右的 dip 分别是 375 → 392 → 440

![auto_size](https://raw.githubusercontent.com/CxmyDev/screen_autosize/main/art/auto_size.jpg)

## Roadmap

- 支持横竖屏切换，自动调整参数；（暂未支持）
- 支持字体尺寸倍数设置，可遵循系统设置，也可自定义；（暂未支持）

## About Me

- Email: plokmju@gmail.com
- 掘金：https://juejin.cn/user/3227821828236392
- 公众号：承香墨影（ID:cxmyDev）

CHANGELOG.md
## [0.0.1+2] 

- fix doc.

## [0.0.1] 

- first version.

## [0.0.1-beta.1] 

- fix doc.

lib/autosize_utils.dart
part of screen_autosize;

class AutoSizeUtils {
  static const double NON_INIT_RATIO = 0.0;

  double baseScreenWidth = 375;

  // 工厂模式
  factory AutoSizeUtils() => instance;

  static AutoSizeUtils get instance => _getInstance();
  static AutoSizeUtils _instance;

  AutoSizeUtils._internal();

  void initConfig({@required double baseWidth}) {
    baseScreenWidth = baseWidth;

    devicePixelRatio = NON_INIT_RATIO;
    _tryInit();
  }

  void _tryInit() {
    if (devicePixelRatio != NON_INIT_RATIO) {
      return;
    }
    var window = WidgetsBinding.instance?.window ?? ui.window;
    devicePixelRatio = window.physicalSize.width / baseScreenWidth;
    mediaWidth = baseScreenWidth;
    mediaHeight = window.physicalSize.height / devicePixelRatio;
    statusBarHeight = window.padding.top / devicePixelRatio;
    bottomBarHeight = window.padding.bottom / devicePixelRatio;
  }

  static AutoSizeUtils _getInstance() {
    if (_instance == null) {
      _instance = new AutoSizeUtils._internal();
    }
    _instance._tryInit();
    return _instance;
  }

  EdgeInsets getPadding() {
    return EdgeInsets.only(top: statusBarHeight, bottom: bottomBarHeight);
  }

  Size get screenSize => Size(mediaWidth, mediaHeight);

  // 屏幕宽
  double mediaWidth;

  // 屏幕高
  double mediaHeight;

  //屏幕像素密度
  double devicePixelRatio = NON_INIT_RATIO;

  double bottomBarHeight;

  double statusBarHeight;
}


example/lib/main_old.dart
import 'package:flutter/material.dart';
import 'package:screen_autosize/screen_autosize.dart';
import 'dart:ui' as ui show window;

void main() => runApp(MyApp());

class Binding extends AutoSizeWidgetsFlutterBinding{
  static WidgetsBinding ensureInitialized() {
    if (WidgetsBinding.instance == null) Binding();
    return WidgetsBinding.instance;
  }
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      // builder: (context, widget) {
      //   return MediaQueryWrapper(builder: (BuildContext context){
      //     return widget;
      //   },);
      // },
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(title: 'Flutter ScreenAutoSize示例'),
    );
  }
}

class ColorsWidget extends StatelessWidget{
  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(10), // 不需要转换，设计稿是多少，就写多少
      width: 100,  // 不需要转换，设计稿是多少，就写多少
      height: 100, // 不需要转换，设计稿是多少，就写多少
      color: Colors.orange,
      child: Text(
        '我是正方形,边长是100',
        style: TextStyle(
          color: Colors.white,
          fontSize: 12,
        ),
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    printScreenInformation();
    var sysMediaData = MediaQueryData.fromWindow(ui.window);
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Row(
              children: <Widget>[
                Container(
                  padding: EdgeInsets.all(10),
                  width: 180,
                  height: 80,
                  color: Colors.red,
                  child: Text(
                    '实际宽度:${180}dp \n'
                        '实际高度:${80}dp',
                    style: TextStyle(color: Colors.white, fontSize: 12),
                  ),
                ),
                Container(
                  padding: EdgeInsets.all(10),
                  width: 180,
                  height: 80,
                  color: Colors.blue,
                  child: Text(
                      '设计稿宽度: 180dp \n'
                          '设计稿高度: 80dp',
                      style: TextStyle(
                          color: Colors.white,
                          fontSize: 12)),
                ),
              ],
            ),
            Container(
              padding: EdgeInsets.all(10),
              width: 375,
              height: 80,
              color: Colors.blueGrey,
              child: Text(
                  '设计稿宽度: 375dp \n'
                      '设计稿高度: 80dp',
                  style: TextStyle(
                      color: Colors.white,
                      fontSize: 12)),
            ),
            Container(
              padding: EdgeInsets.all(10),
              width: MediaQuery.of(context).size.width,
              height: 80,
              color: Colors.green,
              child: Text(
                  '设计稿要求宽度撑满屏幕，使用 MediaQuery.of().size.width 设置 \n'
                      '设计稿高度: 80dp',
                  style: TextStyle(
                      color: Colors.white,
                      fontSize: 12)),
            ),
            Container(
              padding: EdgeInsets.all(10),
              width: 100,
              height: 100,
              color: Colors.orange,
              child: Text(
                '我是正方形,边长是100',
                style: TextStyle(
                  color: Colors.white,
                  fontSize: 12,
                ),
              ),
            ),

            Text('设备原始像素尺寸: ${ui.window.physicalSize.width} x ${ui.window.physicalSize.height} (px)'),
            Text('设备原始 density: ${sysMediaData.devicePixelRatio}dp'),
            Text('设备原始宽度: ${sysMediaData.size.width} dp'),
            Text('设备原始高度: ${sysMediaData.size.height} dp'),
          ],
        ),
      ),
    );
  }

  void printScreenInformation() {
    var sysMediaData = MediaQueryData.fromWindow(ui.window);
    print('设备原始像素尺寸: ${ui.window.physicalSize.width} x ${ui.window.physicalSize.height} (px)');
    print('设备原始 density: ${sysMediaData.devicePixelRatio}dp');
    print('设备原始宽度: ${sysMediaData.size.width} dp');
    print('设备原始高度: ${sysMediaData.size.height} dp');
    print("=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=");
    print('调整后的的像素密度: ${AutoSizeUtils.instance.devicePixelRatio}');
    print('调整后的宽度: ${AutoSizeUtils.instance.mediaWidth} dp');
    print('调整后的高度: ${AutoSizeUtils.instance.mediaHeight} dp');
    print('状态栏高度: ${AutoSizeUtils.instance.statusBarHeight} dp');
    print('底部安全区距离: ${AutoSizeUtils.instance.bottomBarHeight} dp');
  }
}

lib/media_query_wrapper.dart
part of screen_autosize;

class MediaQueryWrapper extends StatelessWidget {

  /// Creates a widget that delegates its build to a callback.
  ///
  /// The [builder] argument must not be null.
  const MediaQueryWrapper({
    Key key,
    @required this.builder,
  }) : assert(builder != null),
        super(key: key);

  /// Called to obtain the child widget.
  ///
  /// This function is called whenever this widget is included in its parent's
  /// build and the old widget (if any) that it synchronizes with has a distinct
  /// object identity. Typically the parent's build method will construct
  /// a new tree of widgets and so a new Builder child will not be [identical]
  /// to the corresponding old one.
  final WidgetBuilder builder;

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
        data: MediaQuery.of(context).copyWith(
          textScaleFactor: 1.0,
          size: AutoSizeUtils.instance.screenSize,
          padding: AutoSizeUtils.instance.getPadding(),
          devicePixelRatio: AutoSizeUtils.instance.devicePixelRatio,
        ),
        child: builder(context));
  }
}

lib/autosize_widgets_flutter_binding.dart
part of screen_autosize;

class AutoSizeWidgetsFlutterBinding extends WidgetsFlutterBinding {

  static WidgetsBinding ensureInitialized() {
    if (WidgetsBinding.instance == null) AutoSizeWidgetsFlutterBinding();
    return WidgetsBinding.instance;
  }

  @override
  ViewConfiguration createViewConfiguration() {
    return ViewConfiguration(
      size: AutoSizeUtils.instance.screenSize,
      devicePixelRatio: AutoSizeUtils.instance.devicePixelRatio,
    );
  }

  /// 因为调整了 PixelRatio，所以对于 Event 事件，需要额外对事件的坐标进行对应比例的转换；
  @override
  void initInstances() {
    super.initInstances();
    ui.window.onPointerDataPacket = _handlePointerDataPacket;
  }

  @override
  void unlocked() {
    super.unlocked();
    _flushPointerEventQueue();
  }

  final Queue<PointerEvent> _pendingPointerEvents = Queue<PointerEvent>();

  void _handlePointerDataPacket(ui.PointerDataPacket packet) {
    // Event 事件，需要按比例转换尺寸
    _pendingPointerEvents.addAll(PointerEventConverter.expand(
        packet.data,
        AutoSizeUtils.instance.devicePixelRatio));
    if (!locked) _flushPointerEventQueue();
  }

  @override
  void cancelPointer(int pointer) {
    if (_pendingPointerEvents.isEmpty && !locked)
      scheduleMicrotask(_flushPointerEventQueue);
    _pendingPointerEvents.addFirst(PointerCancelEvent(pointer: pointer));
  }

  void _flushPointerEventQueue() {
    assert(!locked);
    while (_pendingPointerEvents.isNotEmpty)
      _handlePointerEvent(_pendingPointerEvents.removeFirst());
  }

  final Map<int, HitTestResult> _hitTests = <int, HitTestResult>{};

  void _handlePointerEvent(PointerEvent event) {
    assert(!locked);
    HitTestResult result;
    if (event is PointerDownEvent) {
      assert(!_hitTests.containsKey(event.pointer));
      result = HitTestResult();
      hitTest(result, event.position);
      _hitTests[event.pointer] = result;
    } else if (event is PointerUpEvent || event is PointerCancelEvent) {
      result = _hitTests.remove(event.pointer);
    } else if (event.down) {
      result = _hitTests[event.pointer];
    } else {
      return; // We currently ignore add, remove, and hover move events.
    }
    if (result != null) dispatchEvent(event, result);
  }
}

/**
 * 提供 WidgetsFlutterBinding 的同名方法
 */
void ensureAutoSizeInitialized(){
  AutoSizeWidgetsFlutterBinding.ensureInitialized();
}

/**
 * 替代 runApp()
 */
void runAutoSizeApp(Widget app) {
  var runAppCallback = (){
    AutoSizeWidgetsFlutterBinding.ensureInitialized()
      ..scheduleAttachRootWidget(app)
      ..scheduleWarmUpFrame();
  };
  // 屏幕尺寸初始化较晚，此操作可限制在尺寸赋值后继续操作
  if (window.physicalSize.isEmpty) {
    VoidCallback oldMetricsChanged = window.onMetricsChanged;
    window.onMetricsChanged = () {
      if (!window.physicalSize.isEmpty) {
        if (oldMetricsChanged != null) {
          oldMetricsChanged();
        }
        window.onMetricsChanged = oldMetricsChanged;
        runAppCallback();
      }
    };
  } else {
    runAppCallback();
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:screen_autosize/screen_autosize.dart';
import 'dart:ui' as ui show window;

void main() {
  AutoSizeUtils.instance.initConfig(baseWidth: 375);
  runAutoSizeApp(MyApp());
}

class Binding extends AutoSizeWidgetsFlutterBinding{
  static WidgetsBinding ensureInitialized() {
    if (WidgetsBinding.instance == null) Binding();
    return WidgetsBinding.instance;
  }
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      builder: (context, widget) {
        return MediaQueryWrapper(builder: (BuildContext context){
          return widget;
        },);
      },
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(title: 'Flutter ScreenAutoSize示例'),
    );
  }
}

class ColorsWidget extends StatelessWidget{
  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.all(10), // 不需要转换，设计稿是多少，就写多少
      width: 100,  // 不需要转换，设计稿是多少，就写多少
      height: 100, // 不需要转换，设计稿是多少，就写多少
      color: Colors.orange,
      child: Text(
        '我是正方形,边长是100',
        style: TextStyle(
          color: Colors.white,
          fontSize: 12,
        ),
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    printScreenInformation();
    var sysMediaData = MediaQueryData.fromWindow(ui.window);
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            Container(
              padding: EdgeInsets.only(left: 10),
              width: MediaQuery.of(context).size.width,
              height: AutoSizeUtils.instance.statusBarHeight,
              color: Colors.orange,
              alignment: Alignment.centerLeft,
              child: Text(
                '我和 StatusBar 的高度一致，高度是：${AutoSizeUtils.instance.statusBarHeight} dp',
                style: TextStyle(
                  color: Colors.white,
                  fontSize: 12,
                ),
              ),
            ),
            Row(
              children: <Widget>[
                Container(
                  padding: EdgeInsets.all(10),
                  width: 180,
                  height: 80,
                  color: Colors.red,
                  child: Text(
                    '实际宽度:${180}dp \n'
                        '实际高度:${80}dp',
                    style: TextStyle(color: Colors.white, fontSize: 12),
                  ),
                ),
                Container(
                  padding: EdgeInsets.all(10),
                  width: 180,
                  height: 80,
                  color: Colors.blue,
                  child: Text(
                      '设计稿宽度: 180dp \n'
                          '设计稿高度: 80dp',
                      style: TextStyle(
                          color: Colors.white,
                          fontSize: 12)),
                ),
              ],
            ),
            Container(
              padding: EdgeInsets.all(10),
              width: 375,
              height: 80,
              color: Colors.blueGrey,
              child: Text(
                  '设计稿宽度: 375dp \n'
                      '设计稿高度: 80dp',
                  style: TextStyle(
                      color: Colors.white,
                      fontSize: 12)),
            ),
            Container(
              padding: EdgeInsets.all(10),
              width: MediaQuery.of(context).size.width,
              height: 80,
              color: Colors.green,
              child: Text(
                  '设计稿要求宽度撑满屏幕，使用 MediaQuery.of().size.width 设置 \n'
                      '设计稿高度: 80dp',
                  style: TextStyle(
                      color: Colors.white,
                      fontSize: 12)),
            ),
            Container(
              padding: EdgeInsets.all(10),
              width: 100,
              height: 100,
              color: Colors.orange,
              child: Text(
                '我是正方形,边长是100',
                style: TextStyle(
                  color: Colors.white,
                  fontSize: 12,
                ),
              ),
            ),

            Text('设备原始像素尺寸: ${ui.window.physicalSize.width} x ${ui.window.physicalSize.height} (px)'),
            Text('设备原始 density: ${sysMediaData.devicePixelRatio}dp'),
            Text('设备原始宽度: ${sysMediaData.size.width} dp'),
            Text('设备原始高度: ${sysMediaData.size.height} dp'),
            Text("=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-="),
            Text('调整后的的像素密度: ${AutoSizeUtils.instance.devicePixelRatio}'),
            Text('调整后的宽度: ${AutoSizeUtils.instance.mediaWidth} dp'),
            Text('调整后的高度: ${AutoSizeUtils.instance.mediaHeight} dp'),
            Text('状态栏高度: ${AutoSizeUtils.instance.statusBarHeight} dp'),
            Text('底部安全区距离: ${AutoSizeUtils.instance.bottomBarHeight} dp'),
            Visibility(
              visible: AutoSizeUtils.instance.bottomBarHeight > 0,
              child: Container(
                padding: EdgeInsets.only(left: 10),
                width: MediaQuery.of(context).size.width,
                height: AutoSizeUtils.instance.bottomBarHeight,
                color: Colors.orange,
                alignment: Alignment.centerLeft,
                child: Text(
                  '我和 BottomBar 的高度一致，高度是：${AutoSizeUtils.instance.bottomBarHeight} dp',
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 12,
                  ),
                ),
              ),
            ),

          ],
        ),
      ),
    );
  }

  void printScreenInformation() {
    var sysMediaData = MediaQueryData.fromWindow(ui.window);
    print('设备原始像素尺寸: ${ui.window.physicalSize.width} x ${ui.window.physicalSize.height} (px)');
    print('设备原始 density: ${sysMediaData.devicePixelRatio}dp');
    print('设备原始宽度: ${sysMediaData.size.width} dp');
    print('设备原始高度: ${sysMediaData.size.height} dp');
    print("=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=");
    print('调整后的的像素密度: ${AutoSizeUtils.instance.devicePixelRatio}');
    print('调整后的宽度: ${AutoSizeUtils.instance.mediaWidth} dp');
    print('调整后的高度: ${AutoSizeUtils.instance.mediaHeight} dp');
    print('状态栏高度: ${AutoSizeUtils.instance.statusBarHeight} dp');
    print('底部安全区距离: ${AutoSizeUtils.instance.bottomBarHeight} dp');
  }
}